### PR TITLE
[5.3] Adding display labels feature and improving Login Page / Module consistency [4.2.6 Bugs]

### PR DESCRIPTION
**Pull request by charvimehradu**
Pull Request related to Issue #39641.

### Summary of Changes
**1B:** Changed the "forgot your password?" and "forgot your username?" buttons to a link to show consistency. 
**1C:** Assigned 100% width to the "Log In" button.
**4 :** Assigning the same language strings to all Remember me texts

**1A_2A_2B_3C:** 
These problems also involved the backend files.

- Before, users could choose whether to display icons or words in the login modules. The login module's user icon and the password eye icon image will show up if we choose the option to display labels as icons rather than text. Both the backend login page and the login menu lack this icon. In light of issue 1A, the user icon was thus removed.

- We could display text or icons using the display label option, which was only accessible for the login module. The password hide/show indicator, which is often an important element of the password input fields, disappeared if we choose to display text.
Also when we choose the text option, a heading is added to the input field but the placeholders remain, which has no use.

- In an attempt to make up for the removal of the user icon and with regard to the above point, I also altered the display labels option from text/icon to text/placeholders, allowing the password hide/show icon to always be visible.
As a result, the function was changed, and now, if I select text, only the input field's heading label is shown. If we select placeholders, however, the labels are hidden and only the placeholders are visible.

- I likewise added the same feature and the option of display labels: placeholder/text to the login menu page.

**updating 1A_2A_2B_3C:** 
After some consideration, I realized that icons are a better solution because they provide more clarity and are more universally recognized and also solves the accessibility issues.

### Actual result BEFORE applying this Pull Request

_**1B_1C_4: Login Page Old**_
1B_1C_4_old

_**1A_2A_2B_3C: Login Module Old**_ 
https://user-images.githubusercontent.com/121369234/216763838-76c1a606-cc10-4ea3-9e66-29135a2b40e6.mp4

### Expected result AFTER applying this Pull Request

_**1B_1C_4: Login Page New**_
1B_1C_4

_**1A_2A_2B_3C: Login Module New**_ 
https://user-images.githubusercontent.com/121369234/216763878-f56b1492-b22f-46bd-bd3d-969cd8a7bc67.mp4

_**1A_2A_2B_3C: Login Page New**_ 
https://user-images.githubusercontent.com/121369234/216763884-9b154362-1640-4452-a74e-dc5205d0154f.mp4

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by Kostelano** [comment]
Screenshot_1

When using the **Display Labels: Placeholders**, the distance between the fields is violated (compare with another parameter value) + no field check in the browser - "Please fill in this field".

Screenshot_2

**Comment by brianteeman** [comment]
Not a fan of this

image

image

my 2c is that I don't see the need for this pr - others may disagree

**Comment by charvimehradu** [comment]
@brianteeman If possible, I would like to better understand your concerns about the pull request, and in particular, whether your concern is about the entire issue or just a specific commit.  This would help me to address any concerns you may have and ensure that my work aligns with the goals and objectives of the issue. I value your input and would appreciate any feedback you can provide.
Thank you for your time and consideration. I look forward to the opportunity to continue contributing to the success of Joomla!

**Comment by brianteeman** [comment]
A toggle switch like this really only works when you have two _opposites_ because you can assume the hidden value that will be revealed from the displayed value. eg Yes/No, Show/Hide

https://github.com/joomla/joomla-cms/pull/39787#issuecomment-1417283163

For the component login the option is only available if you create a menu item for _login_ but it is not required for you to create a login menu item for the login to be displayed.

What this means in real terms is that without a menu item the component will use the default values in the view. 

Before this PR that meant the login and password field were displayed with a label and without a placeholder.

With this PR the login and password fields are not displayed at all unless you create a menu item (or if you already have a menu item then you must open and save it again). This is because there is no default value in the view for _login_label_ so the _if_ fails to match either of the conditions.

**Comment by charvimehradu** [comment]
Thank you for bringing this issue to my attention. I understand that the change in the display of the labels and placeholders in the login component may not be what some users expect after updating the software.

The change was made to improve the user experience by streamlining the login process. However, I can see how this change could cause unexpected behavior for some users. As a solution, I propose adding an option to switch between displaying only labels, only placeholders, or both. This way, users can choose the display that works best for their needs.

Would this solution address your concern, I hope I am getting you right and do you have any further suggestions?

**Comment by brianteeman** [comment]
How do any of these changes streamline the process? I don't understand

**Comment by charvimehradu** [comment]
By removing the label and placeholder text, we can reduce visual clutter and make the login process more efficient. If the labels and placeholders displayed the exact same text, then having both visible might not provide much additional value to the user. In such a case, having one or the other (either the label or placeholder) might suffice. Also a potential solution to the issue you've described is to add a default value in the view for the login_label. This way, if there is no menu item created, the toggle switch component would have a default value to fall back on, ensuring that the login and password fields are displayed even if a menu item is not created.

**Comment by drmenzelit** [comment]
You are talking about user experience, but forgot about accessibility... Reducing visual clutter is thinking only about "visual" users. If you only use placeholders you introduce an issue for user with cognitive impairments and they are not always available for assistive technologies.
https://www.w3.org/WAI/GL/low-vision-a11y-tf/wiki/Placeholder_Research

In addition the code is wrong, if you select placeholders the label is still being rendered, but it is empty.

**Comment by charvimehradu** [comment]
Thank you for bringing up the issue with accessibility @drmenzelit . I agree with you now that reducing visual clutter is not a one-size-fits-all solution and accessibility should always be a priority. In regards to the code, I apologize for the mistake. Can you suggest an ideal solution that would address both the visual clarity and accessibility concerns? I would greatly appreciate your insights and expertise on this matter.

**Comment by drmenzelit** [comment]
If you use placeholders, the label should be still present but visually hidden. You can check the login module, there it is correct.

I'm pinging @crystalenka an @angieradtke from UX team, please give your opinion on that.

**Comment by crystalenka** [comment]
Thanks for tagging me, @drmenzelit !

@charvimehradu, I appreciate your attention to these things and your ideas to improve the user experience! I know that we are reviewing your PR pretty thoroughly; please know your work here is appreciated and I hope to see more contributions from you in the future. :)

**Regarding placeholders and labels:**
It's important not to confuse visual clutter with important context. In additional to the points raised by Viviana about people using assistive technology on the web, accessibility also relates to cognitive function as well as sighted but low vision users. While removing visible labels in favor of only placeholders is "cleaner", you lose the context of the field purpose as soon as that placeholder disappears. Additionally, placeholders do not have sufficient contrast to act as labels on their own.

[Here's a good read from Duque about placeholders in forms that goes in a little more detail.](https://www.deque.com/blog/accessible-forms-the-problem-with-placeholders/)

**Regarding the login button:**
In the context of a login page (vs the login module), I think that a start-aligned button is preferred to a full-width button. It makes sense in the module to have the button full width because it's usually shown in a different context, but on a login page that could be quite wide, a full-width button does not make as much sense and impacts the visual hierarchy. I'd be in favor of leaving the button as is, or maybe with a minimum width set so it's not quite so narrow (but not full width either).

**Regarding the styling of the "forgot" links:**
This is subjective but have to be cautious of when this would be merged since it could affect people's templates and/or overrides.

Hope this helps! Thank you. :)

**Comment by angieradtke** [comment]
Dear @charvimehradu, 

we always should  see things in the overall context. Reducing the information presented to the most necessary seems to make sense at first glance.  Because everything that appears in addition must be processed cognitively. But here you start unfortunately from wrong conditions. 

 Joomla has written itself on the flag to be acccessible and I consider a possible configuration of inaccessible display to be wrong. @crystalenka  has summarized the problem quite well. 

We must always look at things from the perspective of all users. It doesn't help people with cognitive disabilities if only the placeholder is visible and disappears when clicked. They might have forgotten what they wanted to type in ...., I might do the same when I'm distracted.

The width of the button is a thing of the template. Everybody can handle it the way he thinks, because this is also always a question in which context the module appears. 

All the best Angie

**Comment by charvimehradu** [comment]
Dear @angieradtke @crystalenka ,

Thank you for reaching out and for your continued support in the project. I understand the importance of accessibility and making sure that the user experience is not compromised. I apologize for the delay in my reply, as I've been focused on preparing for my internal exams.

However, I am here to help and want to make sure that we can come to a solution that benefits everyone. I will take into consideration the points raised about the placeholders and labels, and make sure to provide proper context for all users and thank you for referring me the article on accessible forms from Deque. It was a valuable read and provided a good insight into the problem with using placeholders in forms.

Regarding the login button, I appreciate the explanation about its alignment in different contexts. I will consider adjusting the width to maintain the visual hierarchy, but not make it full-width.

I understand the potential impact that the styling of the "forgot" links could have on templates and overrides. Again, I will be cautious and consider the subjectivity of this aspect while making any changes.

Thank you again for your input. I look forward to continuing to work on improving the user experience while also ensuring accessibility. :)

**Comment by charvimehradu** [comment]
Hi @brianteeman @drmenzelit @crystalenka @angieradtke ,

I just wanted to let you all know that I've made some changes to the pull request. I've reverted back to using icons instead of placeholders because I realized their importance. Additionally, I wanted to clarify that I haven't removed the eye icon for the password field because it's an important feature. I would appreciate your views on whether we should keep the eye icon or not, so please let me know your thoughts. As always, I welcome any feedback or suggestions you may have.

Thanks for your help and support.

File: administrator/language/en-GB/joomla.ini [hunk 1]
```diff
@@ -215,6 +215,9 @@ JFIELD_BASIC_LOGIN_DESCRIPTION_LABEL="Login Description Text"
 JFIELD_BASIC_LOGIN_DESCRIPTION_SHOW_LABEL="Login Description"
 JFIELD_BASIC_LOGOUT_DESCRIPTION_LABEL="Logout Description Text"
 JFIELD_BASIC_LOGOUT_DESCRIPTION_SHOW_LABEL="Logout Description"
+JFIELD_DISPLAY_LABELS="Display Labels"
+JFIELD_DISPLAY_LABELS_TEXT="Text"
+JFIELD_DISPLAY_LABELS_ICONS="Icons"
```
Comment (Quy) [inline administrator/language/en-GB/joomla.ini]: Please sort keys in alpha order.

File: components/com_users/tmpl/login/default.xml [hunk 2]
```diff
@@ -75,6 +75,18 @@
...
+				<field
+					name="login_label"
+					type="radio"
+					label="JFIELD_DISPLAY_LABELS"
+					default="0"
+					layout="joomla.form.field.radio.switcher"
+					filter="integer"
+				>
+				<option value="0">JFIELD_DISPLAY_LABELS_ICONS</option>
+				<option value="1">JFIELD_DISPLAY_LABELS_TEXT</option>
```
Comment (Quy) [inline components/com_users/tmpl/login/default.xml]:
```suggestion
					filter="integer"
					>
					<option value="0">JFIELD_DISPLAY_LABELS_ICONS</option>
					<option value="1">JFIELD_DISPLAY_LABELS_TEXT</option>
```

File: modules/mod_login/tmpl/default.php [hunk 3]
```diff
@@ -58,7 +58,11 @@
                         <span class="visually-hidden"><?php echo Text::_('JSHOWPASSWORD'); ?></span>
                     </button>
                 </div>
+<<<<<<< HEAD
             <?php else : ?>
+=======
+            <?php else : ?> 
+>>>>>>> 6219125d76bc066b7320ad3759ec22ca1b553861
```
Comment (Quy) [inline modules/mod_login/tmpl/default.php]: Please fix.

File: administrator/language/en-GB/joomla.ini (hunk 1 above)
Comment (charvimehradu) [reply administrator/language/en-GB/joomla.ini]: done please check

File: modules/mod_login/tmpl/default.php (hunk 3 above)
Comment (charvimehradu) [reply modules/mod_login/tmpl/default.php]: Done. Please check.

File: modules/mod_login/tmpl/default.php [hunk 4]
```diff
@@ -58,7 +58,7 @@
                         <span class="visually-hidden"><?php echo Text::_('JSHOWPASSWORD'); ?></span>
                     </button>
                 </div>
-            <?php else : ?>
+            <?php else : ?>
```
Comment (Quy) [inline modules/mod_login/tmpl/default.php]:
```suggestion
            <?php else : ?>
```

**Comment by charvimehradu** [comment]
I have taken back the login button width to normal, as suggested in the feedback. I've updated the pull request with this change, and I'd be grateful if you could take another look and let me know if any further changes or updates are needed for this PR.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.0-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.1-dev.

**Comment by crimle** [comment]
I have tested this item :white_check_mark: successfully on 7124211cddcf2cc46a1b8bb444b447f7691b69c5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39787">issues.joomla.org/tracker/joomla-cms/39787</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.3-dev.

**Comment by Hackwar** [comment]
Hello @charvimehradu, we talked about this PR in the maintainers team and unfortunately we decided to not go further with this. Thank you for your work and we hope to see you again as a contributor in the future.

**Pull request closed without merging**